File: code_julia/equilibria/boundary_xz.jl
#=============================================================================
Author: Rayla Kurosaki
GitHub: https://github.com/rkp1503
=============================================================================#

module boundary_xz

function get_equilibria(parameter_values)
    equilibria = get_analytical_solutions(parameter_values)
    r₂₁, r₃₁, p, ϕ₁₂, ϕ₂₁, ϕ₁₃, u₁, u₂, u₃, u₄ = parameter_values
    for (i, equilibrium) in enumerate(equilibria)
        x, y, z = equilibrium
        # if !((x > 0) && (z > 0))
        #     removeat!(equilibria, i)
        # end
        cond1 = (u₄/r₃₁)+(1/ϕ₁₃) > 1
        cond2 = r₃₁ > u₄
        if !all([cond1, cond2])
            deleteat!(equilibria, i)
        end
    end
    return equilibria
end;

function get_stable_equilibria!(equilibria, parameter_values)
    r₂₁, r₃₁, p, ϕ₁₂, ϕ₂₁, ϕ₁₃, u₁, u₂, u₃, u₄ = parameter_values
    for (i, equilibrium) in enumerate(equilibria)
        x, y, z = equilibrium
        cond1 = 1-(u₄/r₃₁) > (1-2*(1-ϕ₁₃*(1-(u₄/r₃₁))))/ϕ₁₃
        cond2 = 1-(u₄/r₃₁) > r₂₁*u₂*(1+ϕ₂₁*(1-ϕ₁₃*(1-(u₄/r₃₁)))^2)/(u₁*(1-p))
        cond3 = 1-(u₄/r₃₁) > (r₃₁-u₄)/(2*r₃₁)
        if !all([cond1, cond2, cond3])
            deleteat!(equilibria, i)
        end
    end
    return nothing
end;

function get_analytical_solutions(parameter_values)
    r₂₁, r₃₁, p, ϕ₁₂, ϕ₂₁, ϕ₁₃, u₁, u₂, u₃, u₄ = parameter_values
    x = 1-ϕ₁₃*(1-(u₄/r₃₁))
    z = 1-(u₄/r₃₁)
    return [(x, 0, z)]
end;
    
end

File: code_python/equilibria/boundary_xz.py
"""
Author: Rayla Kurosaki

GitHub: https://github.com/rkp1503
"""


def get_equilibria(parameter_values):
    equilibria = get_analytical_solutions(parameter_values)
    r_yx, r_zx, p, phi_xy, phi_yx, phi_xz, u_1, u_2, u_3, u_4 = parameter_values
    for (i, equilibrium) in enumerate(equilibria):
        x, y, z = equilibrium
        # if not ((x > 0) and (z > 0)):
        #     del equilibria[i]
        #     pass
        cond1 = (u_4 / r_zx) + (1 / phi_xz) > 1
        cond2 = r_zx > u_4
        if not all([cond1, cond2]):
            del equilibria[i]
            pass
        pass
    return equilibria


def get_analytical_solutions(parameter_values):
    r_yx, r_zx, p, phi_xy, phi_yx, phi_xz, u_1, u_2, u_3, u_4 = parameter_values
    x = 1 - phi_xz * (1 - (u_4 / r_zx))
    z = 1 - (u_4 / r_zx)
    return [(x, 0, z)]


def get_stable_equilibria(equilibria, parameter_values):
    r_yx, r_zx, p, phi_xy, phi_yx, phi_xz, u_1, u_2, u_3, u_4 = parameter_values
    for (i, equilibrium) in enumerate(equilibria):
        x, y, z = equilibrium
        cond1 = 1 - (u_4 / r_zx) > (
                1 - 2 * (1 - phi_xz * (1 - (u_4 / r_zx)))) / phi_xz
        cond2 = 1 - (u_4 / r_zx) > r_yx * u_2 * (
                1 + phi_yx * (1 - phi_xz * (1 - (u_4 / r_zx))) ** 2) / (
                        u_1 * (1 - p))
        cond3 = 1 - (u_4 / r_zx) > (r_zx - u_4) / (2 * r_zx)
        if not all([cond1, cond2, cond3]):
            del equilibria[i]
            pass
        pass
    return equilibria
